[SMTChecker] Basic support for inline assembly that over-approximates its effects.

blishko opened the pull request:
This PR adds basic support in SMTChecker for the analysis of Inline Assembly blocks.
The basic approach over-approximates the effects of the assembly in the following way:

- If any Solidity-level local variable is assigned in the assembly, it is considered as having an unknown value at the end of the assembly block.
- If the block modifies storage (e.g., using `sstore`), all storage variables are considered as having an unknown value at the end of the assembly block.
- If the block modifies memory (e.g., using `mstore`), all memory variables are considered as having an unknown value at the end of the assembly block.
- In all other cases, the variables keep their value from just before the assembly block.

This over-approximation at least eliminates false negatives, as were seen in #10619.
It is a first step towards #10596.

Commit:
[SMTChecker] Basic support for inline assembly using over-approximating analysis

blishko commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,11 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
+//		7737_error,
+//		_inlineAsm.location(),
+//		"Assertion checker does not support inline assembly."
+//	);
```
Do we want to keep this warning, remove it, or replace with something like "Results might be imprecise because of over-approximation caused by inline assembly"?

leonardoalt replied:
We can add the message you suggested when we see we need to reset something, and can omit the message when nothing had to be reset

blishko replied:
I have changed the message of the warning.
Since there will always be something over-approximated (if the assembly block is doing anything), I think it is OK to display it every time.

blishko commented on test/libsolidity/smtCheckerTests/inline_assembly/assembly.sol:
```diff
@@ -0,0 +1,18 @@
+}
+// ----
+// Warning 6328: (272-283): CHC: Assertion violation happens here.\nCounterexample:\n\n\nTransaction trace:\nC.constructor()\nC.g()\n    C.f() -- internal call
+// Warning 6328: (315-327): CHC: Assertion violation happens here.\nCounterexample:\n\n\nTransaction trace:\nC.constructor()\nC.g()\n    C.f() -- internal call\n    C.f() -- internal call
```
This double report of the internal call `C.f()` looks like a bug in CEX reporting.

leonardoalt commented on test/libsolidity/smtCheckerTests/inline_assembly/assembly.sol:
```diff
@@ -0,0 +1,18 @@
+contract C {
+	function f() internal pure returns (bool) {
+		bool b;
+		assembly { b := 1 } // This assignment is overapproximated at the moment, we don't know value of b after the assembly block
```
We can support the assignment when the rhs is a literal or a solidity var

blishko replied:
As agreed, this is more complicated than it seems on the first glance.
If we continue in this direction, it will be a separate PR.

leonardoalt commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -2256,6 +2297,18 @@ void SMTEncoder::resetStateVariables()
+{
+	m_context.resetVariables([&](VariableDeclaration const& _variable) {
+		return _variable.referenceLocation() == VariableDeclaration::Location::Storage || _variable.isStateVariable();
+	});
```
I think this is actually the intended behavior of  `resetStateVariables`. So maybe remove `resetStateVariables` and replace by `resetStorageVariables` wherever it's used?

blishko replied:
As agreed, this will be a separate PR.

leonardoalt commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,11 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
-		_inlineAsm.location(),
-		"Assertion checker does not support inline assembly."
-	);
+	struct AssignedExternalsCollector : public yul::ASTWalker
```
```suggestion
	struct AssignedExternalsCollector: public yul::ASTWalker
```

blishko replied:
Fixed.

leonardoalt commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,11 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
+	struct AssignedExternalsCollector : public yul::ASTWalker
+	{
+		using ExternalReferences = map<yul::Identifier const*, InlineAssemblyAnnotation::ExternalIdentifierInfo>;
+		AssignedExternalsCollector(InlineAssembly const& _inlineAsm) : externalReferences(_inlineAsm.annotation().externalReferences)
```
```suggestion
		AssignedExternalsCollector(InlineAssembly const& _inlineAsm): externalReferences(_inlineAsm.annotation().externalReferences)
```

blishko replied:
Fixed.

leonardoalt commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,11 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
+		set<VariableDeclaration const*> assignedVars;
+
+		using yul::ASTWalker::operator();
+		void operator()(yul::Assignment const& _assignment) {
```
```suggestion
		void operator()(yul::Assignment const& _assignment)
		{
```

blishko replied:
Fixed.

blishko commented on test/libsolidity/smtCheckerTests/inline_assembly/assembly.sol:
```diff
@@ -0,0 +1,24 @@
+// Warning 7737: (103-122): Inline assembly may cause SMTChecker to produce spurious warnings (false positives).
+// Warning 7737: (103-122): Inline assembly may cause SMTChecker to produce spurious warnings (false positives).
+// Warning 7737: (103-122): Inline assembly may cause SMTChecker to produce spurious warnings (false positives).
+// Warning 7737: (103-122): Inline assembly may cause SMTChecker to produce spurious warnings (false positives).
```
Seeing this, I think we need a mechanism in SMTChecker to display some warnings only once.

blishko commented:
@leonardoalt , check if you agree with the error message, or I should revert it back to the original one. As I mentioned, since the inline assembly causes potentially quite a big over-approximations, I would keep displaying the warning.

hrkrshnn commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,10 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
 {
+	struct AssignedExternalsCollector: public yul::ASTWalker
+	{
+		using ExternalReferences = map<yul::Identifier const*, InlineAssemblyAnnotation::ExternalIdentifierInfo>;
```
This seems to be only used once, so do we really need the using?

blishko replied:
Removed the using.

hrkrshnn commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,10 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
+		{
+			auto const& vars = _assignment.variableNames;
+			for (auto const& identifier: vars)
+				if (auto it = externalReferences.find(&identifier); it != externalReferences.end())
```
Can you replace this by `valueOrNullptr` from `libsolutil/CommonData.h`?

blishko replied:
Replaced.

hrkrshnn commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -304,10 +307,49 @@ void SMTEncoder::endVisit(Block const& _block)
+		}
+	};
+
+	auto assignedVars = AssignedExternalsCollector(_inlineAsm).assignedVars;
```
I wonder if we really have to write the class `AssignedExternalCollector`.

There is `Assignments` in `libyul/optimiser/NameCollector.h`, which should give you all `_assignment.variableNames`. Then you could just iterate through the names, without having to write a class. What do you think?

blishko replied:
`NameCollector` seems to be collecting `Yul` variables declared in the assembly, at least that's how I am reading it.
What I need are the Solidity local variables referenced (and more precisely assigned) in the assembly.
So I think `NameCollector` is not the right tool, but I will check if there is not something else, thanks for the tip!

blishko replied:
Sorry, I misread your comment initially,`Assignments` looks useful, thanks for the tip!

hrkrshnn replied:
`Assignments` would only give `YulString`. If you really need `Identifier`, then you might have to write a walker class. So not sure.

blishko replied:
Yeah, `Identifier`s would be much better, since I am then accessing a map with `Identifier *` type of keys. I could work around that, but collecting directly `Identifier`s seems cleaner to me.

hrkrshnn commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -2256,6 +2298,18 @@ void SMTEncoder::resetStateVariables()
 
+void SMTEncoder::resetMemoryVariables()
+{
+	m_context.resetVariables([&](VariableDeclaration const& _variable) { return _variable.referenceLocation() == VariableDeclaration::Location::Memory; });
```
A line break?

blishko replied:
Fixed.

hrkrshnn reviewed (commented):
Looks good.

leonardoalt commented:
Also please rebase

blishko commented:
Addressed the comments and rebased.
I am not sure of this should have a changelog entry or not.
It does eliminate some false negatives, but it is still a very over-approximative analysis.

leonardoalt commented:
I think it should have a Changelog entry at least as a bugfix because it fixes the false negatives

leonardoalt commented:
Some weird exceptions in Yul tests

blishko commented:
> Some weird exceptions in Yul tests

I don't understand these failings, and they happen in completely unrelated tests.

blishko commented:
Added a changelog entry and rebased.

hrkrshnn commented on test/libsolidity/smtCheckerTests/inline_assembly/assembly2.sol:
```diff
@@ -0,0 +1,16 @@
+pragma experimental SMTChecker;
+
+contract C {
+	function f() public pure returns (bool) {
+		bool b;
+		int x = 42;
+		assembly { b := 1 }
+		assert(x == 42); // should hold
+		assert(b); // should hold, but fails due to overapproximation
+		return b;
```
```solidity
	function f() public pure returns (bool) {
		bool b;
		int x = 42;
		assembly { b := 1 }
		b = 5
		assert(x == 42); // should hold
		assert(b); // should hold
		return b;
```
Could you also add a test like this? And a similar one for memory / storage.

blishko replied:
@hrkrshnn : `Error: Type int_const 5 is not implicitly convertible to expected type bool`
Did you mean `b = true;` instead of `b = 5;`?

hrkrshnn replied:
Yes!

blishko replied:
Added.
